core-rust/state-manager/src/limits.rs

```diff
@@ -0,0 +1,177 @@
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+use node_common::config::limits::VertexLimitsConfig;
+use radix_engine::transaction::ExecutionMetrics;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum VertexLimitsExceeded {
+    TransactionsCount,
+    TransactionsSize,
+    ExecutionCostUnitsConsumed,
+    SubstateReadSize,
+    SubstateReadCount,
+    SubstateWriteSize,
+    SubstateWriteCount,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum VertexLimitsAdvanceSuccess {
+    VertexFilled,
+    VertexNotFilled,
+}
+
+pub struct VertexLimitsTracker {
+    pub remaining_transactions_count: usize,
+    pub remaining_transactions_size: usize,
+    pub remaining_execution_cost_units_consumed: usize,
+    pub remaining_substate_read_size: usize,
+    pub remaining_substate_read_count: usize,
+    pub remaining_substate_write_size: usize,
+    pub remaining_substate_write_count: usize,
+}
+
+impl VertexLimitsTracker {
+    pub fn new(config: &VertexLimitsConfig) -> Self {
+        Self {
+            remaining_transactions_count: config.max_total_transactions_count,
+            remaining_transactions_size: config.max_total_transactions_size,
+            remaining_execution_cost_units_consumed: config.max_total_execution_cost_units_consumed,
+            remaining_substate_read_size: config.max_total_substate_read_size,
+            remaining_substate_read_count: config.max_total_substate_read_count,
+            remaining_substate_write_size: config.max_total_substate_write_size,
+            remaining_substate_write_count: config.max_total_substate_write_count,
+        }
+    }
+
+    pub fn check_pre_execution(&self, transaction_size: usize) -> Result<(), VertexLimitsExceeded> {
+        if self.remaining_transactions_count < 1 {
+            return Err(VertexLimitsExceeded::TransactionsCount);
+        }
+
+        if self.remaining_substate_read_size < transaction_size {
```

Comment:
hah, a bit late, but I noticed that you probably meant `self.remaining_transactions_size` here :p

I am fixing this during mu update from develop to https://github.com/radixdlt/babylon-node/pull/519.